//
//  TransitionProcessors.cpp
//
//  This file is intended to be included into other source files.
//  It was split into its own file to allow easier testing.
//
//  Created by Chris Jones on 6/29/17.
//
//

//Transition processing helpers

#include <cassert>
#include "FWCore/Utilities/interface/thread_safety_macros.h"

struct FileResources {
  FileResources(EventProcessor& iEP) : ep_(iEP) {}

  ~FileResources() {
    // See the message in catch clause
    CMS_SA_ALLOW try {
      // Don't try to execute the following sequence of functions twice.
      // If the sequence was already attempted and failed, then do nothing.
      if (!closingSequenceAlreadyFailed_) {
        ep_.respondToCloseInputFile();
        ep_.closeInputFile(cleaningUpAfterException_);
        if (needToCloseOutputFiles_) {
          ep_.endProcessBlock(cleaningUpAfterException_, beginProcessBlockSucceeded_);
          ep_.closeOutputFiles();
        }
      }
    } catch (...) {
      if (cleaningUpAfterException_ or not ep_.setDeferredException(std::current_exception())) {
        std::string message(
            "Another exception was caught while trying to clean up files after the primary fatal exception.");
        ep_.setExceptionMessageFiles(message);
      }
    }
  }

  void normalEnd() { cleaningUpAfterException_ = false; }

  EventProcessor& ep_;
  bool cleaningUpAfterException_ = true;
  bool closingSequenceAlreadyFailed_ = false;
  bool beginProcessBlockSucceeded_ = false;
  bool needToCloseOutputFiles_ = false;
};

struct RunResources {
  RunResources(EventProcessor& iEP, edm::ProcessHistoryID iHist, edm::RunNumber_t iRun) noexcept
      : ep_(iEP), history_(iHist), run_(iRun) {}

  ~RunResources() noexcept {
    // Caught exception is propagated via EventProcessor::setDeferredException()
    CMS_SA_ALLOW try {
      //If we skip empty runs, this would be called conditionally
      ep_.endUnfinishedRun(processHistoryID(),
                           run(),
                           globalTransitionSucceeded_,
                           cleaningUpAfterException_,
                           eventSetupForInstanceSucceeded_);
    } catch (...) {
      if (cleaningUpAfterException_ or not ep_.setDeferredException(std::current_exception())) {
        std::string message(
            "Another exception was caught while trying to clean up runs after the primary fatal exception.");
        ep_.setExceptionMessageRuns(message);
      }
    }
  }

  edm::ProcessHistoryID processHistoryID() const { return history_; }

  edm::RunNumber_t run() const { return run_; }

  void normalEnd() { cleaningUpAfterException_ = false; }

  void succeeded() { success_ = true; }

  EventProcessor& ep_;
  edm::ProcessHistoryID history_;
  edm::RunNumber_t run_;
  bool cleaningUpAfterException_ = true;
  bool success_ = false;
  bool globalTransitionSucceeded_ = false;
  bool eventSetupForInstanceSucceeded_ = false;
};

class LumisInRunProcessor {
public:
  ~LumisInRunProcessor() noexcept { normalEnd(); }

  edm::InputSource::ItemType processLumis(EventProcessor& iEP, std::shared_ptr<RunResources> iRun) {
    currentRun_ = std::move(iRun);
    makeSureLumiEnds_ = true;
    return iEP.processLumis(std::shared_ptr<void>{currentRun_});
  }

  void normalEnd() noexcept {
    // Caught exception is propagated via EventProcessor::setDeferredException()
    CMS_SA_ALLOW try {
      if (makeSureLumiEnds_) {
        makeSureLumiEnds_ = false;
        currentRun_->ep_.endUnfinishedLumi();
      }
    } catch (...) {
      if (not currentRun_->ep_.setDeferredException(std::current_exception())) {
        currentRun_->ep_.setExceptionMessageLumis();
      }
    }

    currentRun_.reset();
  }

private:
  std::shared_ptr<RunResources> currentRun_;
  bool makeSureLumiEnds_ = false;
};

class RunsInFileProcessor {
public:
  edm::InputSource::ItemType processRuns(EventProcessor& iEP) {
    bool finished = false;
    auto nextTransition = edm::InputSource::IsRun;
    do {
      switch (nextTransition) {
        case edm::InputSource::IsRun: {
          processRun(iEP);
          nextTransition = iEP.nextTransitionType();
          break;
        }
        case edm::InputSource::IsLumi: {
          nextTransition = lumis_.processLumis(iEP, currentRun_);
          break;
        }
        default:
          finished = true;
      }
    } while (not finished);
    return nextTransition;
  }

  void normalEnd() {
    lumis_.normalEnd();
    if (currentRun_) {
      currentRun_->normalEnd();
      assert(currentRun_.use_count() == 1);
    }
    currentRun_.reset();
  }

private:
  void processRun(EventProcessor& iEP) {
    auto runID = iEP.nextRunID();
    if ((not currentRun_) or (currentRun_->processHistoryID() != runID.first) or (currentRun_->run() != runID.second)) {
      if (currentRun_) {
        //Both the current run and lumi end here
        lumis_.normalEnd();
        if (edm::InputSource::IsStop == iEP.lastTransitionType()) {
          //an exception happened while processing the end lumi
          return;
        }
        currentRun_->normalEnd();
      }
      currentRun_ = std::make_shared<RunResources>(iEP, runID.first, runID.second);
      iEP.readRun();
      iEP.beginRun(runID.first,
                   runID.second,
                   currentRun_->globalTransitionSucceeded_,
                   currentRun_->eventSetupForInstanceSucceeded_);
      //only if we succeed at beginRun should we run writeRun
      currentRun_->succeeded();
    } else {
      //merge
      iEP.readAndMergeRun();
    }
  }

  std::shared_ptr<RunResources> currentRun_;
  LumisInRunProcessor lumis_;
};

class FilesProcessor {
public:
  explicit FilesProcessor(bool iDoNotMerge) : doNotMerge_(iDoNotMerge) {}

  edm::InputSource::ItemType processFiles(EventProcessor& iEP) {
    bool finished = false;
    auto nextTransition = iEP.nextTransitionType();
    if (nextTransition != edm::InputSource::IsFile)
      return nextTransition;
    do {
      switch (nextTransition) {
        case edm::InputSource::IsFile: {
          processFile(iEP);
          nextTransition = iEP.nextTransitionType();
          break;
        }
        case edm::InputSource::IsRun: {
          nextTransition = runs_.processRuns(iEP);
          break;
        }
        default:
          finished = true;
      }
    } while (not finished);
    runs_.normalEnd();

    return nextTransition;
  }

  void normalEnd() {
    runs_.normalEnd();
    if (filesOpen_) {
      filesOpen_->normalEnd();
      filesOpen_.reset();
    }
  }

private:
  void processFile(EventProcessor& iEP) {
    if (not filesOpen_) {
      readFirstFile(iEP);
    } else {
      if (shouldWeCloseOutput(iEP)) {
        //Need to end this run on the file boundary
        runs_.normalEnd();
        gotoNewInputAndOutputFiles(iEP);
      } else {
        gotoNewInputFile(iEP);
      }
    }
  }

  void readFirstFile(EventProcessor& iEP) {
    iEP.readFile();
    assert(iEP.fileBlockValid());
    iEP.respondToOpenInputFile();

    iEP.openOutputFiles();
    filesOpen_ = std::make_unique<FileResources>(iEP);
    filesOpen_->needToCloseOutputFiles_ = true;

    iEP.beginProcessBlock(filesOpen_->beginProcessBlockSucceeded_);
    iEP.inputProcessBlocks();
  }

  bool shouldWeCloseOutput(EventProcessor& iEP) {
    if (doNotMerge_)
      return true;
    return iEP.shouldWeCloseOutput();
  }

  void gotoNewInputFile(EventProcessor& iEP) {
    iEP.respondToCloseInputFile();
    iEP.closeInputFile(false);

    iEP.readFile();
    if (!iEP.fileBlockValid()) {
      // handle case with last file bad and
      // skipBadFiles true
      return;
    }
    iEP.respondToOpenInputFile();

    iEP.inputProcessBlocks();
  }

  void gotoNewInputAndOutputFiles(EventProcessor& iEP) {
    {
      // If this is still true when we hit the destructor for
      // the filesOpen_ object, then we will know an exception
      // was thrown on one of the following 5 lines.
      filesOpen_->closingSequenceAlreadyFailed_ = true;

      iEP.respondToCloseInputFile();
      bool cleaningUpAfterException = false;
      iEP.closeInputFile(cleaningUpAfterException);
      iEP.endProcessBlock(cleaningUpAfterException, filesOpen_->beginProcessBlockSucceeded_);
      iEP.closeOutputFiles();

      filesOpen_->needToCloseOutputFiles_ = false;
      filesOpen_->closingSequenceAlreadyFailed_ = false;
    }
    {
      filesOpen_->beginProcessBlockSucceeded_ = false;

      iEP.readFile();
      if (!iEP.fileBlockValid()) {
        // handle case with last file bad and
        // skipBadFiles true
        return;
      }
      iEP.respondToOpenInputFile();

      iEP.openOutputFiles();
      filesOpen_->needToCloseOutputFiles_ = true;

      iEP.beginProcessBlock(filesOpen_->beginProcessBlockSucceeded_);
    }
    iEP.inputProcessBlocks();
  }

  std::unique_ptr<FileResources> filesOpen_;
  RunsInFileProcessor runs_;
  bool doNotMerge_;
};
